tempesta_fw/http.c

```diff
@@ -3043,124 +3096,138 @@ do {									\
+			TFW_STR_FOR_EACH_CHUNK(chunk, dup, chunk_end) {
+				if (chunk->flags & TFW_STR_HDR_VALUE)
+					break;
+				hval.nchunks++;
+				hval.len += chunk->len;
+			}
+			r |= tfw_msg_write(&it, &hval);
+			r |= tfw_msg_write(&it, &dlm);
+			hval.chunks += hval.nchunks;
+			hval.nchunks = dup->nchunks - hval.nchunks;
+			hval.len = dup->len - hval.len;
+			r |= tfw_msg_write(&it, &hval);
+
+			r |= tfw_msg_write(&it, &crlf);
+		}
+		if (unlikely(r))
+			goto err;
 	}
-	WRITE_LIT(dst, S_CRLF);
 
-	/*
-	 * We have already added 'host' header, thus, it won't be needed in
-	 * the writing cycle for all headers below.
-	 */
-	TFW_STR_INIT(&ht->tbl[TFW_HTTP_HDR_HOST]);
+	r |= tfw_msg_write(&it, &h_xff);
+	r |= tfw_msg_write(&it, &h_via);
+	/* Finally close headers. */
+	r |= tfw_msg_write(&it, &crlf);
 
-	FOR_EACH_HDR_FIELD_FROM(fld, fld_end, req, TFW_HTTP_HDR_REGULAR) {
-		if (TFW_STR_EMPTY(fld))
-			continue;
-		TFW_STR_FOR_EACH_DUP(hdr, fld, dup_end)
-			WRITE_HDR(dst, hdr);
-	}
+	if (unlikely(r))
+		goto err;
 
-	/*
-	 * Headers, which should be added unconditionally, inserted in
-	 * the end.
-	 */
-	memcpy_fast(__TFW_STR_CH(&h_via, 2)->data, g_vhost->hdr_via,
-		    g_vhost->hdr_via_len);
-	WRITE_STR(dst, &h_via);
-	WRITE_STR(dst, &h_conn);
-	WRITE_LIT(dst, S_CRLF);
+	T_DBG3("%s: req [%p] converted to http1.1\n", __func__, req);
 
-	T_DBG3("%s: req adjusted, it->hb_len=%lu, it->hdrs_len=%lu,"
-	       " it->hdrs_cnt=%u, dst=[%p], pg=[%p]\n", __func__, it->hb_len,
-	       it->hdrs_len, it->hdrs_cnt, dst, (char *)page_address(pg));
+	old_head = req->msg.skb_head;
+	req->msg.skb_head = new_head;
 
-	WARN_ON_ONCE(dst - (char *)page_address(pg) != it->hdrs_len);
+	/* Http chains might add a mark for the message, keep it. */
+	new_head->mark = old_head->mark;
 
-	r = ss_skb_replace_page(&req->msg.skb_head, pg, it->hdrs_len,
-				it->hb_len);
-	if (r) {
-		put_page(pg);
-		return r;
+	if (!TFW_STR_EMPTY(&req->body)) {
+		/*
+		 * Request has a body. we have to detach it from the old
+		 * skb_head and append to a new one. There might be trailing
+		 * headers after the body, but we're already copied them before
+		 * body. This is not a problem, but we have to drop the trailer
+		 * part after the body to avoid sending the same headers twice.
+		 *
+		 * Body travels in a separate DATA frame thus it's always in
+		 * it's own skb.
+		 */
+		struct sk_buff *b_skbs = old_head, *trailer;
+		size_t len = 0;
+
+		do {
+			b_skbs = b_skbs->next;
+			if (WARN_ON_ONCE((b_skbs == old_head)))
+				goto err;
+		} while (b_skbs != req->body.skb);
+
+		ss_skb_queue_split(old_head, b_skbs);
+		trailer = b_skbs;
+		do {
+			len += trailer->len;
+			trailer = trailer->next;
+
+		} while ((trailer != b_skbs) || (len != req->body.len));
+		ss_skb_queue_tail(&req->msg.skb_head, b_skbs);
+		if (trailer != b_skbs) {
+			ss_skb_queue_split(req->msg.skb_head, trailer);
+			ss_skb_queue_tail(&old_head, trailer);
+		}
 	}
+	ss_skb_queue_purge(&old_head);
```

Comment:
The `ss_skb_replace_page()` function is not used any more, and looks like it can be removed.